OpenRA.Mods.Common/Traits/Mobile.cs

```diff
@@ -145,234 +50,20 @@ IEnumerable<object> IActorPreviewInitInfo.ActorPreviewInits(ActorInfo ai, ActorP
-			{
-				Speed = speed;
-				Cost = cost;
-			}
-		}
-
-		public struct WorldMovementInfo
-		{
-			internal readonly World World;
-			internal readonly TerrainInfo[] TerrainInfos;
-			internal WorldMovementInfo(World world, MobileInfo info)
-			{
-				// PERF: This struct allows us to cache the terrain info for the tileset used by the world.
-				// This allows us to speed up some performance-sensitive pathfinding calculations.
-				World = world;
-				TerrainInfos = info.TilesetTerrainInfo[world.Map.Rules.TileSet];
-			}
-		}
-
-		public readonly Cache<TileSet, TerrainInfo[]> TilesetTerrainInfo;
-		public readonly Cache<TileSet, int> TilesetMovementClass;
-
-		public MobileInfo()
-		{
-			TilesetTerrainInfo = new Cache<TileSet, TerrainInfo[]>(LoadTilesetSpeeds);
-			TilesetMovementClass = new Cache<TileSet, int>(CalculateTilesetMovementClass);
-		}
-
-		public int MovementCostForCell(World world, CPos cell)
-		{
-			return MovementCostForCell(world, TilesetTerrainInfo[world.Map.Rules.TileSet], cell);
-		}
-
-		int MovementCostForCell(World world, TerrainInfo[] terrainInfos, CPos cell)
-		{
-			if (!world.Map.Contains(cell))
-				return int.MaxValue;
-
-			var index = cell.Layer == 0 ? world.Map.GetTerrainIndex(cell) :
-				world.GetCustomMovementLayers()[cell.Layer].GetTerrainIndex(cell);
-
-			if (index == byte.MaxValue)
-				return int.MaxValue;
-
-			return terrainInfos[index].Cost;
-		}
-
-		public int CalculateTilesetMovementClass(TileSet tileset)
-		{
-			// collect our ability to cross *all* terraintypes, in a bitvector
-			return TilesetTerrainInfo[tileset].Select(ti => ti.Cost < int.MaxValue).ToBits();
-		}
-
-		public int GetMovementClass(TileSet tileset)
-		{
-			return TilesetMovementClass[tileset];
-		}
-
-		static bool IsMovingInMyDirection(Actor self, Actor other)
-		{
-			var otherMobile = other.TraitOrDefault<Mobile>();
-			if (otherMobile == null || !otherMobile.IsMoving)
-				return false;
-
-			var selfMobile = self.TraitOrDefault<Mobile>();
-			if (selfMobile == null)
-				return false;
-
-			// Moving in the same direction if the facing delta is between +/- 90 degrees
-			var delta = Util.NormalizeFacing(otherMobile.Facing - selfMobile.Facing);
-			return delta < 64 || delta > 192;
-		}
-
-		public int TileSetMovementHash(TileSet tileSet)
-		{
-			var terrainInfos = TilesetTerrainInfo[tileSet];
-
-			// Compute and return the hash using aggregate
-			return terrainInfos.Aggregate(terrainInfos.Length,
-				(current, terrainInfo) => unchecked(current * 31 + terrainInfo.Cost));
-		}
+		public int GetInitialFacing() { return InitialFacing; }
 
+		public bool SharesCell = false;
 		public bool CanEnterCell(World world, Actor self, CPos cell, Actor ignoreActor = null, bool checkTransientActors = true)
 		{
-			if (MovementCostForCell(world, cell) == int.MaxValue)
+			var movementClassInfo = world.WorldActor.TraitsImplementing<MovementClass>().Single(mc => mc.Info.Class == MovementClass).Info;
```

Comment:
The worldActor does not itself include MovementClass Trait, which the mobile instances are looking for.

Are you sure a trait on the world the right mechanism for obtaining an instance?

Seems like you want something of a instance management mechanism in relation to the world actor, to request instantiation of the MovementClass, if it doesn't already exist(the world probably shouldn't need to declare what specific Movement Class instances are needed).

Reply:
OpenRA uses traits on the world to hold various types of global state, including palettes, faction definitions, and references for the in-game UI.  Defining the movement types on the world also matches up with the original games (which called them "locomotors"), and makes it more explicit that these are things that should be limited in number.

Reply:
But please do this as `world.WorldActor.Info.TraitInfos<MovementClassInfo>().Single(mc => mc.Class == MovementClass)` instead of going via the actual world actor.